parcel-bundler/parcel
# Implement profiling with a keyboard shortcut

This implements the ability to profile parcel once it's running and capture only a segment of the run (without capturing the whole run using `--profile`), using a keyboard shortcut when stdin is a tty.

It's currently bound to ~`ctrl-y`~`ctrl-e` and toggles capturing a profile whether or not Parcel was invoked with `--profile`. If it was, it ends the profile. If it wasn't, it starts a profile and a subsequent press will end it. If a profile hasn't ended before Parcel ends, Parcel will end the profile itself.

This also exposes public api on the main `Parcel` object for `startProfiling()`, `stopProfiling()` to start and stop profiling respectively, and `isProfiling` to query its state. The methods support reentrance and throw if the expected state mismatches (e.g. profiling is requested once profiling has already started).

Test Plan: 
* Run the kitchen sink build with and without `--profile`. In each case, press ~`ctrl-y`~`ctrl-e` early and verify that it toggles state and, if applicable, writes a profile. In the case that a profile continues to run, verify Parcel ends the profile when it completes the run.
* Verify ctrl-c results in a non-zero exit code
* Verify watch mode continues to function, and profiling works there as well
